File: src/test_stack.py
"""Test if the stack stays consistent."""
from build._rbtree_tests import lib, ffi
from hypothesis import settings
from hypothesis.stateful import GenericStateMachine
from hypothesis.strategies import tuples, just, integers


class Item(object):
    """Represent a item in the stack."""

    def __init__(self, item, value):
        item.value = value
        self.item = item


def test_stackx():
    """Test if stack works as expected."""
    lib.test_stack_init()
    stack = []
    for x in [1, 2, 3, 4]:
        item = Item(ffi.new("item_t*"), x)
        stack.append(item)
        assert lib.test_push(item.item) == 0
    assert lib.test_pop(4) == 0
    assert lib.test_stack_iter([3, 2, 1]) == 0
    assert lib.test_pop(3) == 0


class GenStack(GenericStateMachine):
    """Test if stack the stays consistent."""

    def __init__(self):
        self.comparison = []
        lib.test_stack_init()

    def steps(self):
        push_item = tuples(
            just("push"),
            integers(
                min_value=-2**30,
                max_value=(2**30) - 1
            )
        )
        pop_item = tuples(just(
            "pop"
        ), just(None))
        if not self.comparison:
            return push_item
        else:
            return (
                push_item | pop_item
            )

    def execute_step(self, step):
        action, value = step
        if action == 'push':
            item = Item(ffi.new("item_t*"), value)
            self.comparison.append(item)
            assert lib.test_push(item.item) == 0
            lib.test_stack_iter([
                item.item.value for item in reversed(self.comparison)
            ])
        elif action == 'pop':
            item = self.comparison.pop()
            assert lib.test_pop(item.item.value) == 0
            assert item.item.next == ffi.NULL
            if not self.comparison:
                assert lib.test_stack_null() == 0
        else:
            assert False


# with settings(max_examples=2000):
TestStack = GenStack.TestCase

File: src/test_stack.c
#include "testing.h"

item_t* stack = NULL;

void
test_stack_init(void)
{
    stack = NULL;
}

int
test_push(item_t* item)
{
    qs_push(&stack, item);
    return 0;
}

int
test_pop(int item)
{
    item_t *tmp = NULL;
    qs_pop(&stack, &tmp);
    TA(item == rb_value_m(tmp), "Wrong item popped");
    return 0;
}

int
test_stack_null(void)
{
    return stack != NULL;
}

int
test_stack_iter(int* values)
{
    (void)(values);
    qs_stack_iter_decl_cx_m(qq, iter, elem);
    int i = 0;
    // TODO can we make iteration type safe?
    rb_for_m(qs, stack, iter, elem) {
        if(rb_value_m(elem) != values[i])
            return 1;
        i += 1;
    }
    return 0;
}
